flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/TtlStateFactory.java

```diff
@@ -137,32 +139,41 @@ private IS createState() throws Exception {
     @SuppressWarnings("unchecked")
     private IS createValueState() throws Exception {
         ValueStateDescriptor<TtlValue<SV>> ttlDescriptor =
-                new ValueStateDescriptor<>(
-                        stateDesc.getName(),
-                        new TtlSerializer<>(LongSerializer.INSTANCE, stateDesc.getSerializer()));
+                stateDesc.getSerializer() instanceof TtlSerializer
```

Comment:
If not introducing changelog state backend, will this problem still existed?

Reply:
Not in Heap and RocksDB as far as I know.